torch/csrc/distributed/c10d/TCPStore.cpp

```diff
@@ -350,22 +356,50 @@ TCPStore::TCPStore(std::string host, const TCPStoreOptions& opts)
     addr_.port = opts.port;
   }
 
-  if (numWorkers_.has_value()) {
-    std::random_device rd;
-    std::mt19937 gen(rd());
-    std::uniform_int_distribution<> distrib(1, *numWorkers_);
-    // TODO (xilunwu): this wait logic may be removed after fixing read_offset
-    // stagger connecting to the store when there are too many ranks to
-    // avoid causing a DDoS
-    std::this_thread::sleep_for(std::chrono::milliseconds(distrib(gen)));
-  }
+  // Try connecting several times -- if the server listen backlog is full it may
+  // fail on the first send in validate.
+  auto deadline = std::chrono::steady_clock::now() + opts.timeout;
+  auto backoff = std::make_shared<ExponentialBackoff>();
+
+  auto retry = 0;
+  do {
+    try {
+      client_ = detail::TCPClient::connect(addr_, opts, backoff);
+      // TCP connection established
+      C10D_DEBUG("TCP client connected to host {}:{}", addr_.host, addr_.port);
+
+      // client's first query for validation
+      validate();
+
+      // success
+      break;
+    } catch (const c10::DistNetworkError& ex) {
+      if (deadline < std::chrono::steady_clock::now()) {
```

Comment:
I wonder if the backoff object can "know" about the deadline

Reply:
I thought about it but decided against it -- but maybe that's the cleaner implementation. Could also wrap the entire retry block in the backoff instead

Still a bit on the fence

Reply:
I think I'm going to keep it as is -- abstracting away some of this logic would be nice but I'm not sure there's a clean way to report errors etc in the way we want if we don't own the loop